src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -594,7 +555,7 @@ private bool TryUpdateInternal(TKey key, int? nullableHashcode, TValue newValue,
                         {
                             if (valueComparer.Equals(node._value, comparisonValue))
                             {
-                                if (s_isValueWriteAtomic)
+                                if (!typeof(TValue).IsValueType || ConcurrentDictionaryTypeProps<TValue>.IsWriteAtomic)
```

Comment:
It's not intuitive for someone reading the code why the IsValueType check is necessary here, given IsWriteAtomic includes it.  I'd suggest pulling this check out into an aggressively inlined method or property that includes a comment as to why this is written exactly the way it is (similar to your PR comments), and then use that prop in both places it's needed.